crates/ruff_python_formatter/src/other/arguments.rs

```diff
@@ -1,12 +1,188 @@
+            kwonlyargs,
+            kw_defaults,
+            kwarg,
+        } = item;
+
+        let saved_level = f.context().node_level();
+        f.context_mut().set_node_level(NodeLevel::Expression);
+
+        let format_inner = format_with(|f: &mut PyFormatter| {
+            let separator = format_with(|f| write!(f, [text(","), soft_line_break_or_space()]));
+            let mut joiner = f.join_with(separator);
+            let mut last_node: Option<AnyNodeRef> = None;
+
+            let mut defaults = std::iter::repeat(None)
+                .take(posonlyargs.len() + args.len() - defaults.len())
+                .chain(defaults.iter().map(Some));
+
+            for positional in posonlyargs {
+                let default = defaults.next().ok_or(FormatError::SyntaxError)?;
+                joiner.entry(&ArgumentWithDefault {
+                    argument: positional,
+                    default,
+                });
+
+                last_node = Some(default.map_or_else(|| positional.into(), AnyNodeRef::from));
+            }
+
+            if !posonlyargs.is_empty() {
+                joiner.entry(&text("/"));
+            }
+
+            for argument in args {
+                let default = defaults.next().ok_or(FormatError::SyntaxError)?;
+
+                joiner.entry(&ArgumentWithDefault { argument, default });
+
+                last_node = Some(default.map_or_else(|| argument.into(), AnyNodeRef::from));
+            }
+
+            if let Some(vararg) = vararg {
+                joiner.entry(&format_args![
+                    leading_node_comments(vararg.as_ref()),
+                    text("*"),
+                    vararg.format().with_options(ArgumentKind::Varg)
+                ]);
+                last_node = Some(vararg.as_any_node_ref());
+            }
+
+            debug_assert!(defaults.next().is_none());
+
+            let mut defaults = std::iter::repeat(None)
+                .take(kwonlyargs.len() - kw_defaults.len())
+                .chain(kw_defaults.iter().map(Some));
+
+            for keyword_argument in kwonlyargs {
+                let default = defaults.next().ok_or(FormatError::SyntaxError)?;
+                joiner.entry(&ArgumentWithDefault {
+                    argument: keyword_argument,
+                    default,
+                });
+
+                last_node = Some(default.map_or_else(|| keyword_argument.into(), AnyNodeRef::from));
+            }
+
+            debug_assert!(defaults.next().is_none());
+
+            if let Some(kwarg) = kwarg {
+                joiner.entry(&format_args![
+                    leading_node_comments(kwarg.as_ref()),
+                    text("**"),
+                    kwarg.format().with_options(ArgumentKind::Kwarg)
+                ]);
+                last_node = Some(kwarg.as_any_node_ref());
+            }
+
+            joiner.finish()?;
+
+            write!(f, [if_group_breaks(&text(","))])?;
+
+            if let Some(last_node) = last_node {
```

Comment:
```suggestion
            // Break group if there is a trailing comma
            if let Some(last_node) = last_node {
```

Reply:
Does this also need to read the magic trailing comma setting?

Reply:
If we introduce a setting for this, then yes

Reply:
i forgot this isn't merged: https://github.com/astral-sh/ruff/pull/4878/files#diff-c1f1071e384c9f14936cad67780f301345744791e34afd93cada2cf43af8dcb5R18

I haven't looked into how to pass settings, but i think it makes sense to introduce the boolean for that now

Reply:
The way this works is that Ruff should define a `PyFormatOption` struct that implements `FormatOptions` and is passed to the `PyFormatContext`. You can then access it with `f.context().options()....`

https://github.com/charliermarsh/ruff/blob/dc7cdb04e17cd9aae7ab63976424b01bf040f8d3/crates/ruff_formatter/src/lib.rs#L257-L277